packages/typescript-client/src/client.ts

```diff
@@ -207,53 +212,60 @@ export class ShapeStream {
       }
 
       try {
-        await this.fetchClient(url.toString(), { signal })
-          .then(async (response) => {
-            if (!response.ok) {
-              throw await FetchError.fromResponse(response, url.toString())
-            }
+        const response = await this.fetchClient(url.toString(), { signal })
 
-            const { headers, status } = response
-            const shapeId = headers.get(`X-Electric-Shape-Id`)
-            if (shapeId) {
-              this.shapeId = shapeId
-            }
+        if (!response.ok) {
+          throw await FetchError.fromResponse(response, url.toString())
+        }
 
-            const lastOffset = headers.get(`X-Electric-Chunk-Last-Offset`)
-            if (lastOffset) {
-              this.lastOffset = lastOffset as Offset
-            }
+        const { headers, status } = response
+        const shapeId = headers.get(`X-Electric-Shape-Id`)
+        if (shapeId) {
+          this.shapeId = shapeId
+        }
+
+        const lastOffset = headers.get(`X-Electric-Chunk-Last-Offset`)
+        if (lastOffset) {
+          this.lastOffset = lastOffset as Offset
+        }
 
-            attempt = 0
+        const schemaHeader = headers.get(`X-Electric-Schema`)!
+        this.schema = schemaHeader ? JSON.parse(schemaHeader) : {}
 
-            if (status === 204) {
-              return []
-            }
+        attempt = 0
 
-            return response.json() as Promise<Message[]>
-          })
-          .then((batch: Message[]) => {
-            // Update isUpToDate & lastOffset
-            if (batch.length > 0) {
-              const lastMessages = batch.slice(-2)
+        const messages = status === 204 ? `[]` : await response.text()
+        const batch = JSON.parse(messages, (key, value) => {
+          if (key === `value`) {
+            // Parse the row values
+            const row = value as Record<string, Value>
+            Object.keys(row).forEach((key) => {
+              row[key] = this.parse(key, row[key] as string)
+            })
+          }
+          return value
+        }) as Message[]
 
-              lastMessages.forEach((message) => {
-                if (message.headers?.[`control`] === `up-to-date`) {
-                  const wasUpToDate = this.isUpToDate
+        // Update isUpToDate & lastOffset
+        if (batch.length > 0) {
+          const lastMessages = batch.slice(-2)
```

Comment:
since you're refactoring, this no longer updates `lastOffset` at all, given we're doing it from headers now. So we can just check the last element for up-to-date instead of slicing last 2 and iterating over them,=